federator: rename Brig -> Service and add galley

pcapriotti opened the pull request:
Federator is currently forwarding all its requests to brig, regardless of the value of the `Component` field. This PR adds support for arbitrary internal services, renames the `Brig` effect to `Service` to reflect its generalised function, and adds a configuration option for galley.

jschaul reviewed (commented):
Hm, interesting! We didn't catch this before, when we only use mocks... but it would be nice to be able to catch this sort of problem without an end2end test.

As https://github.com/wireapp/wire-server/blob/develop/services/brig/test/integration/Federation/End2end.hs#L236 returns a 200, if galley can't be called yet, I suppose that is because currently only a call to a remote brig is made, and https://github.com/wireapp/wire-server/pull/1556 isn't merged yet? 

(Even though this is only a draft, changes so far look good)

pcapriotti commented:
> Hm, interesting! We didn't catch this before, when we only use mocks... but it would be nice to be able to catch this sort of problem without an end2end test.

Agreed, but I'm not sure how.

> As https://github.com/wireapp/wire-server/blob/develop/services/brig/test/integration/Federation/End2end.hs#L236 returns a 200, if galley can't be called yet, I suppose that is because currently only a call to a remote brig is made, and #1556 isn't merged yet?

That's right, that test is indeed failing in #1556.

mdimjasevic commented on services/federator/src/Federator/Run.hs:
```diff
@@ -86,8 +87,8 @@ newEnv o _dnsResolver = do
   let _runSettings = Opt.optSettings o
-  let _brig = mkEndpoint (Opt.brig o)
-  let _brigEndpoint = Opt.brig o
+  let _brig Brig = mkEndpoint (Opt.brig o)
+      _brig Galley = mkEndpoint (Opt.galley o)
```
How about calling this something other than `_brig` as it is not about Brig only anymore? Perhaps `_componentEndpoint` or something like that?

pcapriotti replied:
Good point, I forgot to name the field.

mdimjasevic commented on services/federator/src/Federator/Service.hs:
```diff
@@ -32,30 +30,31 @@ import Federator.Env (brig)
+interpretService = interpret $ \case
+  ServiceCall component path body domain -> embed @Federator . liftAppIOToFederator $ do
     brigReq <- view brig <$> ask
     res <-
-      rpc' "brig" brigReq $
+      rpc' "brig" (brigReq component) $
         RPC.method HTTP.POST
           . RPC.path path -- FUTUREWORK(federation): Protect against arbitrary paths
           . RPC.body (RPC.RequestBodyBS body)
```
I do not understand this part, but I see that renaming/generalizing in the PR from a brig to a service did not change this section much. This is not a Brig-specific request anymore, right? So perhaps `serviceReq` instead of `brigReq`? In the same vein, that `"brig"` string seems overly specific; I am not sure what exactly that argument to `rpc'` represents, but my understanding so far is that it is used only in printing/showing an exception, so how about using something like `show component` instead?

mdimjasevic commented on services/federator/test/unit/Test/Federator/ExternalServer.hs:
```diff
@@ -48,27 +48,27 @@ tests =
-    runM . evalMock @Brig @IO $ do
-      mockBrigCallReturns @IO (\_ _ _ -> pure (HTTP.notFound404, Just "response body"))
+    runM . evalMock @Service @IO $ do
+      mockServiceCallReturns @IO (\_ _ _ _ -> pure (HTTP.notFound404, Just "response body"))
```
Where does this `mockServiceCallReturns` thing come from? I can't find it anywhere in the codebase.

pcapriotti replied:
It's automatically generated by `genMock`.

mdimjasevic commented on services/federator/test/unit/Test/Federator/ExternalServer.hs:
```diff
@@ -48,27 +48,27 @@ tests =
 
-      actualCalls <- mockBrigCallCalls @IO
-      let expectedCall = ("/users", "\"foo\"", aValidDomain)
+      actualCalls <- mockServiceCallCalls @IO
```
Can't find `mockServiceCallCalls` either.

mdimjasevic commented on services/federator/test/unit/Test/Federator/ExternalServer.hs:
```diff
@@ -36,7 +36,7 @@ import Test.Tasty (TestTree, testGroup)
+genMock ''Service
 
 tests :: TestTree
 tests =
```
Perhaps at some point it would make sense to add tests for Galley as well. If that makes sense, I'm fine with doing that in a separate PR.

pcapriotti replied:
I added one test for galley.

mdimjasevic reviewed (changes requested):
Overall this looks good. The only thing I am not sure about is the implementation of `interpretService`. @pcapriotti , can you tell if it needs updating?

pcapriotti commented:
Thanks for the review. I changed whatever `brig` name was left into `service`. I also updated the label passed to `rpc` with the actual name of the service, and added a test for galley requests.

mdimjasevic commented on services/federator/test/unit/Test/Federator/ExternalServer.hs:
```diff
@@ -72,6 +73,19 @@ requestBrigFailure =
+
+      res :: InwardResponse <- mock @Service @IO . noLogs . Polysemy.runReader allowAllSettings $ callLocal request
+      actualCalls <- mockServiceCallCalls @IO
+      let expectedCall = (Galley, "/users", "\"foo\"", aValidDomain)
```
I don't understand this well, but shouldn't this be about a Galley-related endpoint?

pcapriotti replied:
I guess it should. It's just a mocked request, so it doesn't matter, but I'll turn it into a galley endpoint for clarity.

mdimjasevic reviewed (approved):
Great!

I left a tiny comment in a Galley test you added.
